### PR TITLE
refactor(experimental): add cluster types to rpc-transport

### PR DESCRIPTION
**Pull request by buffalojoec**

### CODE REVIEW

**Comment by buffalojoec** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #1970** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1970?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #1971** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1971?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1972** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1972?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #1973** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1973?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #1974** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1974?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1974?utm_source=stack-comment).

**Comment by lorisleiva** [comment]
I needed a moment to let that simmer but I think it's an API that makes sense.

You're essentially preparing a set of RPC methods and flagging them as devnet, testnet or mainnet compatible.

Two questions:

- What happens when we have a set of RPC methods available on all clusters? Should we mark them as `IRpcApiMethodsDevnet<T> & IRpcApiMethodsTestnet<T> & IRpcApiMethodsMainnet<T>`? If so, with the current implementation, the RPC returned will always be an `RpcDevnet` since it's the first TypeScript condition in the return type.
- Do we want to worry about other clusters like `localhost` that are a lot more custom?

**Comment by buffalojoec** [comment]
> * What happens when we have a set of RPC methods available on all clusters? Should we mark them as `IRpcApiMethodsDevnet<T> & IRpcApiMethodsTestnet<T> & IRpcApiMethodsMainnet<T>`? If so, with the current implementation, the RPC returned will always be an `RpcDevnet` since it's the first TypeScript condition in the return type.

Yeah, good point. My thought was that one should instantiate a set of types exactly like I've done in the [type tests](packages/rpc-transport/src/__typetests__/methods-api-typetest.ts). But I realize that people could very easily end up doing what you're describing. Maybe there's a better way to declare an RPC or API instance as `devnet | testnet | mainnet`? 

I toyed with the idea of providing just a string arg for the cluster, but wasn't sure how idiomatic I felt it was.

> * Do we want to worry about other clusters like `localhost` that are a lot more custom?

Yeah I think it might make sense to include a localhost one.

File: packages/rpc-transport/src/apis/api-types.ts [hunk 1]
```diff
@@ -11,6 +11,10 @@ export interface IRpcApiMethods {
     [methodName: string]: RpcMethod;
 }
 
+export type IRpcApiMethodsDevnet<TRpcMethods> = TRpcMethods & { readonly __devnet: true };
+export type IRpcApiMethodsTestnet<TRpcMethods> = TRpcMethods & { readonly __testnet: true };
+export type IRpcApiMethodsMainnet<TRpcMethods> = TRpcMethods & { readonly __mainnet: true };
```
[the remark this answers is not in the document]
Comment (steveluscher) [reply packages/rpc-transport/src/apis/api-types.ts:16]: Hey @buffalojoec, does [this help at all](https://www.typescriptlang.org/play?ssl=28&ssc=1&pln=29&pc=1#code/JYOwLgpgTgZghgYwgAgEoQI4FcIGcwCCwUAJlAPYAOyA3gFDKPIDkAfpPiBGMwFzJgoOANwMmbALZxQXHv3gAbXBFFMWrEhABusvgKEqxjNgvII4C3f0EijyKJhz4ipCpQAUYcgGsIIXPwgWBIARtAAlPz4UKAA5qIAvnR0oJCwiCgA4twAwuTkCgAqABZxtHZsHGBW+rZqktIgNTaG9RrazQaq4qym5pbcei3djLG5+UWlILHukchBodAjyGNgeQUlce4IEwDyWtCmcCTWBnPRcYnJYACelCgElMAAstzF5CS4yAC8yNlrE0202QAB80I48IRiGQqKI6Dt-GBkHAfvMsAoFMivlgQN4QOQAO4gLHIR4vN4fXAAIRuOQUWHw0AAPGTXmB3p8ADQsKQyQYAPjhCPwyBCqKCGJJOLxhOJcC+rIpnxpdIZaRZTzZHNw3OYfQsukFyWFSIQ4vRmPlyGl+KJJMV7MpKvpjKgGvJjq5LCqhqF5ERyBI5slVptsvtmqV1NpLvVDu1us0OgFfoDKF+Est2NxtrlCsjntwAFEAB5IShgZ1q5nxym63lNFPJAD0rbb7Y7na7neudxQqtdqL1ZgNg1BLCTunHzB9Y7BzAbujht3upKehT7uHuCAAqiBgP7CuQAJLgaDKBBgA8gJk7-k-OzuHfICAlyAgT7IkA3ZAAfmQ7glvwuwSMAYC3rqrCLnO6j6gMPDTu0yYIfO7CQoa4Q-PeYZ2oE2gRC+b5+J+7iAfwoAwNAyDHph3zYTmsrhHY-7HnYeEHFAy59muHralWrpMoUtZeoUA5pIR76fmJ0D3t8dhkhu9xbhAu77oeJ5nlAF5Xv6TJ2GoomxlRr6SV8C6NLo+lMP+paCIg4FCQWCa0Oo0FyLUKAJPyVmMPwhnVlAEnEWZs48D5f7ILZUD2YJwk6i5lToYMpw4MgXnhX50mBSZwUTh0gzhTZb7RZesVOXWCVIZ0qXpWoaiZUZ2VER+ZlwZZdXWZFxUxY5vEVTQsEjvBQwGGl3kdb58z4VAdhGnQK4POVnyluWlYxgFZV9SJWVBS1yBZbJ8nrpu257teR6nmk2nXnpE3IP5g45Xt5l8mFd1FXZpW9Vq-WuRZyXIIoyhjRl907U9n4zklb0TR9JUOXF3IDYlnAA0DnnjRNDUBbtkOTgV71dZ9CNLfFyNVWjFjA7VWNg41uOtUN7Ww0T8ObT9Xrk21lNKBjoNcBxs2iEAA)?

Essentially it's a way to declare _on the method spec itself_ what clusters it applies to, and then filter them anywhere by cluster.

**Comment by steveluscher** [comment]
Idea for TypeScript magic: https://github.com/solana-labs/solana-web3.js/pull/1974#discussion_r1431991541

File: packages/rpc-transport/src/apis/api-types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (buffalojoec) [reply packages/rpc-transport/src/apis/api-types.ts:16]: Yeah I think this definitely helps, but I might need to massage it a little bit. I noticed you can still get away with:

```ts
const f = null as unknown as ApiMethodsExceptByCluster<RequestAirdrop, 'mainnet'>;
f.requestAirdrop(1)
```

However, I think adding method-level cluster definitions makes sense, even though it's going to be quite verbose. The tilde and the extracts are very nice touches.

Let me see what I can put together.

[the remark this answers is not in the document]
Comment (buffalojoec) [reply packages/rpc-transport/src/apis/api-types.ts:16]: Oh wait, I think I was supposed to use `ApiMethodsByCluster`. Disregard.

**Comment by buffalojoec** [comment]
FYI I'm doing some stack shuffling and will re-introduce this PR and it's children either here or in a different stack when I come back to this work.

**Pull request closed without merging**